src/lib/reasoners/domains.ml

```diff
@@ -530,8 +530,12 @@ struct
             D.intersect (D.unknown (X.type_info repr)) @@
             D.add_explanation ~ex (Entry.domain entry)
 
-        let set_domain { entry ; explanation = ex ; _ } d =
-          Entry.set_domain entry (D.add_explanation ~ex d)
+        let set_domain { repr ; entry ; explanation = ex } d =
+          if Explanation.is_empty ex then Entry.set_domain entry d
```

Comment:
Why you don't need to intersect the domain `d` if the explanation is empty?

Reply:
We suppose that the domain is within the type bounds originally. The issue we are trying to fix is that `D.add_explanation` discards the global bounds (this is necessary for soundness, because the global bounds are not necessarily type bounds). If the explanation is empty, we don't need to call `add_explanation` (`D.add_explanation` is a no-op in this case), so we don't need to enforce the type bounds.

Reply:
> We suppose that the domain is within the type bounds originally

Now that I think about it, I don't think this is documented, but it should.

Reply:
Clarified the doc about this.